File: zig/pyenc-legacy/ptype.zig
//! Python bindings for the pyenc library.
const std = @import("std");
const py = @import("pydust");
const enc = @import("pyenc");
const pyenc = @import("./pyenc.zig");

pub const PType = py.class(struct {
    const Self = @This();

    wrapped: enc.PType,

    id: py.property(struct {
        pub fn get(self: *const Self) u8 {
            return @intFromEnum(self.wrapped);
        }
    }) = .{},

    bit_width: py.property(struct {
        pub fn get(self: *const Self) u8 {
            return self.wrapped.bitSizeOf();
        }
    }) = .{},

    byte_width: py.property(struct {
        pub fn get(self: *const Self) u8 {
            return self.wrapped.sizeOf();
        }
    }) = .{},

    pub fn __eq__(self: *const Self, other: *const Self) bool {
        return @intFromEnum(self.wrapped) == @intFromEnum(other.wrapped);
    }

    pub fn __hash__(self: *const Self) usize {
        var hasher = std.hash.Wyhash.init(0);
        std.hash.autoHash(&hasher, @intFromEnum(self.wrapped));
        return hasher.final();
    }

    pub fn __repr__(self: *const Self) !py.PyString {
        // Defer to the Zig representation of PType.
        return try py.PyString.createFmt("{}", .{self.wrapped});
    }

    /// Convert the pyenc PType to a pyenc DType.
    pub fn to_dtype(self: *const Self) !*const pyenc.DType {
        return pyenc.DType.wrapOwned(enc.DType.fromPType(self.wrapped));
    }

    /// Convert the pyenc PType to a PyArrow DType.
    pub fn to_pyarrow(self: *const Self) !py.PyObject {
        const pa_dtype = switch (self.wrapped) {
            .i8 => "int8",
            .u8 => "uint8",
            .i16 => "int16",
            .u16 => "uint16",
            .i32 => "int32",
            .u32 => "uint32",
            .i64 => "int64",
            .u64 => "uint64",
            .f16 => "float16",
            .f32 => "float32",
            .f64 => "float64",
        };
        const pa = try py.import("pyarrow");
        defer pa.decref();

        return pa.call0(py.PyObject, pa_dtype);
    }

    /// Convert the pyenc PType to a PyArrow DType.
    pub fn to_format_code(self: *const Self) !py.PyString {
        const format_str = switch (self.wrapped) {
            .i8 => "b",
            .u8 => "B",
            .i16 => "h",
            .u16 => "H",
            .i32 => "i",
            .u32 => "I",
            .i64 => "l",
            .u64 => "L",
            .f16 => "e",
            .f32 => "f",
            .f64 => "d",
        };
        return py.PyString.create(format_str);
    }

    pub usingnamespace py.zig(struct {
        pub fn wrap(ptype: enc.PType) !*const Self {
            return py.init(Self, .{ .wrapped = ptype });
        }

        pub fn unwrap(self: *const Self) enc.PType {
            return self.wrapped;
        }
    });
});

File: vortex-array/src/serde/ptype.rs
use num_enum::{IntoPrimitive, TryFromPrimitive};

use crate::ptype::PType;

#[derive(IntoPrimitive, TryFromPrimitive)]
#[repr(u8)]
pub enum PTypeTag {
    U8,
    U16,
    U32,
    U64,
    I8,
    I16,
    I32,
    I64,
    F16,
    F32,
    F64,
}

impl From<PType> for PTypeTag {
    fn from(value: PType) -> Self {
        match value {
            PType::U8 => PTypeTag::U8,
            PType::U16 => PTypeTag::U16,
            PType::U32 => PTypeTag::U32,
            PType::U64 => PTypeTag::U64,
            PType::I8 => PTypeTag::I8,
            PType::I16 => PTypeTag::I16,
            PType::I32 => PTypeTag::I32,
            PType::I64 => PTypeTag::I64,
            PType::F16 => PTypeTag::F16,
            PType::F32 => PTypeTag::F32,
            PType::F64 => PTypeTag::F64,
        }
    }
}

impl From<PTypeTag> for PType {
    fn from(value: PTypeTag) -> Self {
        match value {
            PTypeTag::U8 => PType::U8,
            PTypeTag::U16 => PType::U16,
            PTypeTag::U32 => PType::U32,
            PTypeTag::U64 => PType::U64,
            PTypeTag::I8 => PType::I8,
            PTypeTag::I16 => PType::I16,
            PTypeTag::I32 => PType::I32,
            PTypeTag::I64 => PType::I64,
            PTypeTag::F16 => PType::F16,
            PTypeTag::F32 => PType::F32,
            PTypeTag::F64 => PType::F64,
        }
    }
}
